File: paper_artifacts/experiments/llr8/artifacts/sources/qwen38-c-skills/fuse_move_ifs__submitted.c
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <omp.h>

void fuse_move_ifs_fp64(
    double *restrict a,
    double *restrict b,
    const double *restrict cond,
    const double *restrict src,
    const int64_t K,
    const int64_t LEN_2D,
    unsigned char *restrict workspace,
    const int64_t workspace_size)
{
    (void)workspace;
    (void)workspace_size;
    const int64_t n = LEN_2D;
    if (n <= 0) return;

    if (K > 0) {
        /* Fused single pass over src: rows with cond[i]>0 also feed a.
           Row axis is independent (no dependence between rows) -> thread it;
           column axis is unit stride and disjoint (restrict) -> vectorize it. */
        #pragma omp parallel for schedule(static)
        for (int64_t i = 0; i < n; i++) {
            const double *restrict s  = src + i * n;
            double *restrict ap = a + i * n;
            double *restrict bp = b + i * n;
            if (cond[i] > 0.0) {
                for (int64_t j = 0; j < n; j++) {
                    ap[j] = s[j] * 2.0;
                    bp[j] = s[j] + 1.0;
                }
            } else {
                for (int64_t j = 0; j < n; j++) {
                    bp[j] = s[j] + 1.0;
                }
            }
        }
    } else {
        #pragma omp parallel for schedule(static)
        for (int64_t i = 0; i < n; i++) {
            if (cond[i] > 0.0) {
                const double *restrict s  = src + i * n;
                double *restrict ap = a + i * n;
                for (int64_t j = 0; j < n; j++) {
                    ap[j] = s[j] * 2.0;
                }
            }
        }
    }
}

File: paper_artifacts/experiments/llr8/artifacts/sources/kimi27sglang-fortran-skills/fuse_move_ifs__submitted.f90
subroutine fuse_move_ifs_fp64(a, b, cond, src, K, LEN_2D, workspace, workspace_size) bind(C)
  use iso_c_binding
  implicit none
  integer(c_int64_t), value, intent(in) :: K, LEN_2D, workspace_size
  type(c_ptr), value, intent(in) :: workspace
  real(c_double), intent(inout) :: a(LEN_2D, LEN_2D), b(LEN_2D, LEN_2D)
  real(c_double), intent(in) :: cond(LEN_2D), src(LEN_2D, LEN_2D)
  integer(c_int64_t) :: i, j
  if (K > 0) then
    !$omp parallel do private(j)
    do i = 1, LEN_2D
      if (cond(i) > 0.0d0) then
        do j = 1, LEN_2D
          a(j, i) = src(j, i) * 2.0d0
        end do
      end if
      do j = 1, LEN_2D
        b(j, i) = src(j, i) + 1.0d0
      end do
    end do
  else
    !$omp parallel do private(j)
    do i = 1, LEN_2D
      if (cond(i) > 0.0d0) then
        do j = 1, LEN_2D
          a(j, i) = src(j, i) * 2.0d0
        end do
      end if
    end do
  end if
end subroutine fuse_move_ifs_fp64
